src/core/thread/time_sync_service.cpp

```diff
@@ -151,6 +154,82 @@ void TimeSync::ProcessTimeSync(void)
 }
 #endif // OPENTHREAD_FTD
 
+void TimeSync::HandleStateChanged(otChangedFlags aFlags)
+{
+    if ((aFlags & OT_CHANGED_THREAD_ROLE) != 0)
+    {
+        CheckAndHandleChanges(false);
+    }
+}
+
+void TimeSync::HandleTimeout()
+{
+    CheckAndHandleChanges(false);
+}
+
+void TimeSync::HandleStateChanged(Notifier::Callback &aCallback, otChangedFlags aFlags)
+{
+    aCallback.GetOwner<TimeSync>().HandleStateChanged(aFlags);
+}
+
+void TimeSync::HandleTimeout(Timer &aTimer)
+{
+    aTimer.GetOwner<TimeSync>().HandleTimeout();
+}
+
+void TimeSync::CheckAndHandleChanges(bool aTimeUpdated)
+{
+    otNetworkTimeStatus networkTimeStatus       = OT_NETWORK_TIME_SYNCHRONIZED;
+    const otDeviceRole  role                    = GetInstance().GetThreadNetif().GetMle().GetRole();
+    const uint32_t      resyncNeededThresholdMs = 2 * TimerMilli::SecToMsec(mTimeSyncPeriod);
+    const uint32_t      timeSyncLastSyncMs      = TimerMilli::GetNow() - mLastTimeSyncReceived;
+    uint32_t            timerTimeoutMs;
+
+    switch (role)
+    {
+    case OT_DEVICE_ROLE_DISABLED:
+    case OT_DEVICE_ROLE_DETACHED:
+        networkTimeStatus = OT_NETWORK_TIME_UNSYNCHRONIZED;
+        break;
+
+    case OT_DEVICE_ROLE_CHILD:
+    case OT_DEVICE_ROLE_ROUTER:
+        if (timeSyncLastSyncMs > resyncNeededThresholdMs)
+        {
+            // The device hasn’t received time sync for more than two periods time.
+            networkTimeStatus = OT_NETWORK_TIME_RESYNC_NEEDED;
+        }
+        break;
+
+    case OT_DEVICE_ROLE_LEADER:
+        break;
+    }
+
+    mTimer.Stop();
+
+    if (networkTimeStatus == OT_NETWORK_TIME_SYNCHRONIZED)
+    {
+        // Schedule a check 1 millisecond after two periods of time
+        if (resyncNeededThresholdMs >= timeSyncLastSyncMs)
+        {
+            timerTimeoutMs = resyncNeededThresholdMs - timeSyncLastSyncMs + 1;
+        }
+        else
+        {
+            timerTimeoutMs = resyncNeededThresholdMs + 1;
```

Comment:
In which case will it run into the code logic here? I think only when the role is Leader, but for Leader we don't need to start the check timer.

Could we simplify the logic as follows:
```
    if (networkTimeStatus == OT_NETWORK_TIME_SYNCHRONIZED && role != OT_DEVICE_ROLE_LEADER)
    {
         // Schedule a check 1 millisecond after two periods of time
         timerTimeoutMs = resyncNeededThresholdMs - timeSyncLastSyncMs + 1
         mTimer.Start(timerTimeoutMs);
    }
```

Reply:
Yes it was because of the leader case leaving `mLastTimeSyncReceived` as 0 but you are right in pointing out we don't need the timer for the leader anyway. We also don't need the timer for disabled and detached, so I've combined the logic in to the switch statement. Furthermore, I added an assert `assert(resyncNeededThresholdMs >= timeSyncLastSyncMs)` because the conditions for preventing this condition are quite subtle and may sting a contributor if they make modifications to the behaviour of time sync in the future.